### PR TITLE
Improve isolation of query cache entries under re-created users or role switches

### LINKED ISSUE #58054: Flush cached result when user switch role
Cache query results are tied to the user and their queries. The authorisation on the data being accessed is checked at the time the queries execute but not when retrieving the cache. As such, Any changes on user's permission in between queries will not be reflected on the cached query result.

Simple steps to test

```
CREATE USER admin IDENTIFIED BY  'password';

GRANT CURRENT GRANTS ON *.* TO admin WITH GRANT OPTION;

-- Login as admin.  Could be done anytime after admin creation, but must be
-- done before table can be queried, after grants are given

-- 1. Create the table
CREATE TABLE user_data
(
    ID UInt32,
    userID UInt32
)
ENGINE = MergeTree
ORDER BY userID;

-- 2. Create roles with row-level security

-- Role for users with userID = 1

CREATE ROLE user_role_1;
GRANT SELECT ON user_data TO user_role_1;
CREATE ROW POLICY user_policy_1 ON user_data
    FOR SELECT USING userID = 1 TO user_role_1;

-- Role for users with userID = 2

CREATE ROLE user_role_2;
GRANT SELECT ON user_data TO user_role_2;
CREATE ROW POLICY user_policy_2 ON user_data
    FOR SELECT USING userID = 2 TO user_role_2;

-- 3. Grant roles to the admin user

GRANT user_role_1, user_role_2 TO admin;

INSERT INTO user_data (ID, userID) VALUES (1, 1), (2, 2), (3, 1), (4, 3), (5, 2), (6, 1), (7, 4), (8, 2);


-- Switch to admin if not already done so

-- Policy test

SET ROLE user_role_1;

-- only returns rows for user_role_1
SELECT * FROM user_data;

SET ROLE ALL;
SET ROLE user_role_2;

-- only returns rows for user_role_2
SELECT * FROM user_data;

SET ROLE ALL;
SET ROLE user_role_1;

-- only returns rows for user_role_1
SELECT * FROM user_data SETTINGS use_query_cache=1;

SET ROLE ALL;
SET ROLE user_role_2;

-- 🚨🚨🚨🚨 user_role_1 ROWS is returned here instead of user_role_2🚨🚨🚨🚨
SELECT * FROM user_data SETTINGS use_query_cache=1;
```

While i think the current behaviour is somewhat expected from a caching layer (this is similar to how web server caching behaves where CDN doesnt have context of user permissions and one would rely on the cache to eventually expire) and the issue only surface in a very particular setup & usage of cached query. For now, i [updated the documentation to warn about this behavior](https://github.com/ClickHouse/ClickHouse/pull/58053) until we solve it properly.

**Describe the solution you'd like**

A simple fix would be to flush any cache result when the roles are updated.

A more complex fix that may help when users switching roles/context all the time and still want to use cached query results is to incorporate into the query cache some sort of hash key with current grant (or at least the rolenames) to be checked when returning cache result.=

### PR DESCRIPTION
**Pull request by rschu1ze**
Fixes #58054

### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
The query cache now denies access to entries when the user is re-created or assumes another role. This improves prevents attacks where 1. an user with the same name as a dropped user may access the old user's cache entries or 2. a user with a different role may access cache entries of a role with a different row policy.

### CODE REVIEW

**Comment by robot-ch-test-poll** [comment]
*This is an automated comment for commit fabc06995e37d77d56bc505c814c43c49a9e3da5 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/58611/fabc06995e37d77d56bc505c814c43c49a9e3da5/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/58611/fabc06995e37d77d56bc505c814c43c49a9e3da5/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

File: src/Interpreters/Cache/QueryCache.h [hunk 1]
```diff
@@ -51,8 +51,16 @@ class QueryCache
...
 
-        /// The user who executed the query.
+        /// The name, id and current roles of the user who executed the query.
+        /// These members are necessary to ensure that a (non-shared, see below) entry can only be written and read by the same user with
+        /// the same roles. Example attack scenarios:
+        /// - after DROP USER, it must not be possible to create a new user with with the dropped user name and access the dropped user's
+        ///   query cache entries
+        /// - different roles of the same user may be tied to different row-level policies. It must not be possible to switch role and
+        ///   access another role's cache entries
         const String user_name;
```
Comment (vitlibar) [inline src/Interpreters/Cache/QueryCache.h:61]: Is `user_name` still used?

Comment (rschu1ze) [reply src/Interpreters/Cache/QueryCache.h:61]: You are right. But before fixing this: do you know why the `user_id` / `getUserID()` in `Context` is an `std::optional`? When would the user id be not set?

**Comment by rschu1ze** [comment]
@vitlibar Approve? Thanks!

File: src/Common/TTLCachePolicy.h [hunk 2]
```diff
@@ -11,37 +13,47 @@ namespace DB
 class PerUserTTLCachePolicyUserQuota : public ICachePolicyUserQuota
 {
 public:
-    void setQuotaForUser(const String & user_name, size_t max_size_in_bytes, size_t max_entries) override
+    void setQuotaForUser(std::optional<UUID> user_id, size_t max_size_in_bytes, size_t max_entries) override
```
Comment (vitlibar) [inline src/Common/TTLCachePolicy.h:16]: `setQuotaForUser()` doesn't look good when its parameter is `std::optional<UUID>`. I mean logically `setQuotaForUser(nullopt, ...)` looks a bit pointless (you set something for an absent user). The function doesn't do anything useful anyway if this parameter is nullopt, so what I suggest is just moving this condition `if (!user_id)` outside of the function and turn the type of the parameter into a plain `const UUID & user_id`. And for other functions `increaseActual()`, `decreaseActual()`, `approveWrite()` it's the same.

File: tests/queries/0_stateless/02494_query_cache_user_isolation.sh [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+# shellcheck source=../shell_config.sh
+. "$CURDIR"/../shell_config.sh
+
+# -- Attack 1:
+#    - create a user,
+#    - run a query whose result is stored in the query cache,
+#    - drop the user, recreate it with the same name
+#    - test that the cache entry is inaccessible
+
+${CLICKHOUSE_CLIENT} --query "SELECT 'Attack 1'"
```
Comment (vitlibar) [inline tests/queries/0_stateless/02494_query_cache_user_isolation.sh:16]: `echo "Attack 1"` is shorter

Comment (rschu1ze) [reply tests/queries/0_stateless/02494_query_cache_user_isolation.sh:16]: Done

**Review by vitlibar — APPROVED** [review]

**Pull request merged**